File: rover.go
package mars_rover

import (
	"fmt"
)

const (
	MaxHeight = 10
	MaxWidth  = 10

	Right = 'R'
	Left  = 'L'
	Move  = 'M'
)

type Rover struct {
	direction  *Direction
	coordinate *Coordinate
}

func NewRover() *Rover {
	return &Rover{
		direction:  NewDirection(),
		coordinate: NewCoordinate(0, 0),
	}
}

func (r *Rover) navigate(commands string) {
	for _, command := range commands {
		switch command {
		case Right:
			r.direction = r.direction.Right()
		case Left:
			r.direction = r.direction.Left()
		case Move:
			r.coordinate = r.coordinate.Next(r.direction)
		}
	}
}

func (r *Rover) position() string {
	return fmt.Sprintf("%d:%d:%s",
		r.coordinate.X,
		r.coordinate.Y,
		r.direction.Value)
}

File: f#/MarsRover.Tests/Rover.fs
namespace MarsRover.Tests

open System

module Rover =
    type Direction =
        | North = 0
        | West = 1
        | South = 2
        | East = 3

    let Left (direction: Direction) =
        match direction with
        | Direction.North -> Direction.West
        | Direction.West -> Direction.South
        | Direction.South -> Direction.East
        | _ -> direction

    type Rover = { direction: Direction }
    let NewRover () = { direction = Direction.North }

    let rotateLeft (rover: Rover) =
        { rover with
              direction = Left(rover.direction) }

    let Position (rover: Rover) =
        match rover.direction with
        | Direction.North -> "0:0:N"
        | Direction.West -> "0:0:W"
        | Direction.South -> "0:0:S"
        | Direction.East -> "0:0:E"
        | _ -> ArgumentOutOfRangeException() |> raise

    let rec Navigate (rover: Rover, command: string) =
        if command.Length = 0 then
            rover
        else
            Navigate(rotateLeft rover, command.[1..])
